File: DxVisualizer/IdxPool.cpp
#include "IdxPool.h"

namespace CoriumDirectX {

    unsigned int IdxPool::acquireIdx() {
        if (recycleList.size() > 0) {
            unsigned int res = recycleList.top();
            recycleList.pop();
            return res;
        }
        else
            return (++highestAcquiredIdx);
    }

    void IdxPool::releaseIdx(int idx) { recycleList.push(idx); }

}

File: Corium3DGI/Utils/IdxPool.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Corium3DGI.Utils
{
    public class IdxPool
    {
        Stack<int> recycleList = new Stack<int>();
        int highestAcquiredIdx = -1;

        public int acquireIdx()
        {
            if (recycleList.Count > 0)
                return recycleList.Pop();
            else
                return (++highestAcquiredIdx);
            
        } 

        public void releaseIdx(int idx)
        {
            verifyReleasedIdx(idx);

            recycleList.Push(idx);
        }

        [Conditional("DEBUG")]
        private void verifyReleasedIdx(int idx)
        {
            if (recycleList.Contains(idx) || idx > highestAcquiredIdx)
                throw new Exception("Invalid index for release : " + idx);
        }
    }
}
